fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,141 @@
+     * update state with the received todos.
+     */
+
+    fetch('./api/todos', { method: 'GET' })
+      .then(res => res.json())
+      .then(data => {
+        this.setState({
+          todos: data
+        });
+      })
+      .catch(error => {
+        console.error('Error:', error);
+      });
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+
+    fetch('./api/todos', {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(newTodo)
+    })
+      .then(res => res.json())
+      .then(data => {
+        this.setState(prevState => {
+          return { todos: prevState.todos.concat(data) };
+        });
+      })
+      .catch(error => {
+        console.error('Error:', error);
+      });
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     * replace the old todo in the state array.
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list should "toggle" its isCompleted status back and forth.
+     *
+     * Do not mutate the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    let isCompletedOpp = {};
+    for (let i = 0; i < this.state.todos.length; i++) {
+      if (todoId === this.state.todos[i].todoId) {
+        if (this.state.todos[i].isCompleted === true) {
+          isCompletedOpp = { isCompleted: false };
+          fetch(`/api/todos/${todoId}`, {
+            method: 'PATCH',
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify(isCompletedOpp)
+          })
+            .then(res => res.json())
+            .then(data => {
+              this.setState(prevState => {
+                const newArray = this.state.todos;
+                newArray.splice(i, 1, data);
```

Comment:
Careful, this is not a copy. Objects (including arrays) are pass-by-reference. However, you can make a copy with the `Array.prototyp.slice()` method.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice